### PR TITLE
Stream api

### PR DESCRIPTION
**Pull request by danron**
First draft of a Stream API. 

It is incomplete. No error handling or parameterisation of buffer size and some missing functions like readBytes(). And some refactoring and tests are needed.

How can we make this work in the clustered solution?

### CODE REVIEW

**Comment by na--** [comment]
Take a look at the [github.com/spf13/afero](afero) library and how we use it elsewhere in k6. Generally we want to avoid direct filesystem calls like `os.Open()`. That's because of multiple reasons: testability of the code, ability to run both script files and archives, security in the cloud execution scenario, etc.

**Comment by na--** [comment]
Take a look at the [initcontext](https://github.com/loadimpact/k6/blob/master/js/initcontext.go) and the `Open()` function in it. Opening local files for streaming should be done in a similar manner and again only accessible in the init context for the reasons listed [here](https://docs.k6.io/docs/running-k6#section-the-init-context-and-the-default-function):
> There are two reasons for this. The first is, of course: performance.
>
> If you read a file from disk on every single script iteration, it'd be needlessly slow; even if you cache the contents of the file and any imported modules, it'd mean the first run of the script would be much slower than all the others. Worse yet, if you have a script that imports or loads things based on things that can only be known at runtime, you'd get slow iterations thrown in every time you load something new.
>
> But there's another, more interesting reason. By forcing all imports and file reads into the init context, we design for distributed execution. We know which files will be needed, so we distribute only those files. We know which modules will be imported, so we can bundle them up from the get-go. And, tying into the performance point above, the other nodes don't even need writable filesystems - everything can be kept in-memory.

I'd also add "node security" to those reasons, at least when it comes to executing scripts in a cluster or in the loadimpact cloud.

Also, notice the usage of [`afero.Fs`](https://godoc.org/github.com/spf13/afero#Fs) there. That's what allows us to plug any one of its many implementations and work with either the real OS filesystem, an in-memory filesystem (very useful for tests) or even do [very easy caching](https://github.com/loadimpact/k6/blob/master/js/bundle.go#L66-L70). We [use](https://github.com/loadimpact/k6/search?utf8=%E2%9C%93&q=afero.New&type=) that flexibility very much in k6.

**Comment by danron** [comment]
If initcontext opens this file it will happen for every VU which is exactly what I am trying to avoid, less memory usage. 

I know the "init context" README by heart by now and I don't agree with it. A global buffered read from file won't slow down things for 99% of the users. It is taking performance to it's extremes way of thinking. 

And how is opening the file for each VU any better?

**Comment by na--** [comment]
Yeah, good point. But we still need to know what are the files that a script depends on, so we can bundle them in the archive when we execute a test in the cluster or cloud mode. That's currently done by tracking which files are opened in the init context.

For normal files, at the moment when you call `open()`, k6 [checks](https://github.com/loadimpact/k6/blob/master/js/initcontext.go#L165-L166) if the file had already been loaded. If you run the following script:
```js
let file = open("./test.txt")

export default function () {
  console.log("test.txt contents: " + file)
}
```

with multiple VUs and iterations and with debug output enabled (for example `k6 run -q -v --vus 2 --iterations 5 script.js`), you'd see that `Loading...                                    name=./test.txt` is present only once in the log. 

The problem with this approach and why it won't work with streams is that it would break if you decide to open the same file twice for streaming. So some architectural refactoring is definitely needed, I'm just not sure what the best way to do it is.

File: js/streams/streams.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+func (fs *FileStream) readCSVLine() []string {
+	out, err := fs.csv.Read()
+	if err == io.EOF {
+		if fs.loop {
+			fs.reset(0)
+		}
+	} else if err != nil {
+		panic(err)
```
Comment (luizbafilho) [inline js/streams/streams.go]: I'm not sure about this panic here. Exploding a stack trace in user's face is never a good ux. Maybe should we return the error instead?

Comment (danron) [reply js/streams/streams.go]: The PR is not complete. So far only a suggestion :)

Comment (danron) [reply js/streams/streams.go]: Look at the big picture stuff so far. We need to solve the streaming as a whole or we will soon have a fork of K6 :P

**Comment by danron** [comment]
I splitted the streaming and the csv part into an initcontext openStream and a csv module. Look at the sample to see how it's used. The interesting change is the streams.Stream interface.

**Comment by MStoykov** [comment]
Thank you for the PR, @danron,  but as has been mentioned here and in #592 this change will not play well with future cluster implementation and the current cloud execution. 

As such and the fact that there has been no movement I am closing this PR. Maybe in the future, around the implementation of cluster execution, we can reuse some of the code and the ideas from here.

**Pull request closed without merging**